git/checkignore.go

```diff
@@ -0,0 +1,244 @@
+package git
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+)
+
+type CheckIgnoreOptions struct {
+	NoIndex        bool
+	CheckSubModule bool
+}
+
+type IgnoreMatch struct {
+	PathName File
+	Pattern  string
+	Source   File
+	LineNum  int
+}
+
+func (im IgnoreMatch) String() string {
+	lineNum := ""
+	if im.LineNum != 0 {
+		lineNum = strconv.Itoa(im.LineNum)
+	}
+	return fmt.Sprintf("%s:%v:%s\t%s", im.Source, lineNum, im.Pattern, im.PathName)
+}
+
+func CheckIgnore(c *Client, opts CheckIgnoreOptions, paths []File) ([]IgnoreMatch, error) {
+	patternMatches := make([]IgnoreMatch, len(paths))
+
+	for idx, path := range paths {
+		if !opts.NoIndex {
+			log.Printf("Checking if %s is tracked by git\n", path.String())
+			entries, err := LsFiles(c, LsFilesOptions{Cached: true, ExcludeStandard: false}, []File{path})
+			if err != nil {
+				return nil, err
+			}
+
+			// As a default, nothing that is tracked by git is ignored
+			if len(entries) > 0 {
+				log.Printf("Path %v is tracked by git and not ignored.\n", path)
+				continue
+			}
+		}
+
+		abs, err := filepath.Abs(path.String())
+		if err != nil {
+			return nil, err
+		}
+
+		// Does this go through a symbolic link?
+		evalPath, _ := filepath.EvalSymlinks(filepath.Dir(abs))
+		log.Printf("Eval symlink path of %v is %v\n", abs, evalPath)
+		if evalPath != "" && evalPath != filepath.Dir(abs) {
+			return nil, fmt.Errorf("fatal: pathspec '%v' is beyond a symbolic link", path)
+		}
+
+		// Does it go through a submodule?
+		if opts.CheckSubModule {
+			submodule := findSubmodule(c, abs)
+			if submodule != "" {
+				submoduleRel, err := filepath.Rel(c.WorkDir.String(), submodule)
+				if err != nil {
+					return nil, err
+				}
+				if submoduleRel != "" {
+					return nil, fmt.Errorf("fatal: Pathspec '%v' is in submodule '%v'", path, submoduleRel)
+				}
+			}
+		}
+
+		stat, _ := os.Lstat(abs)
+		isDir := false
+		if stat != nil {
+			log.Printf("Path %v has an lstat\n", abs)
+			isDir = stat.IsDir()
+		}
+
+		// Let's check that this path is in the git work dir first
+		wdpath, err := filepath.Rel(c.WorkDir.String(), abs)
+		if err != nil || wdpath == "." {
+			return nil, fmt.Errorf("Path %v is not in the git work directory.", path.String())
+		}
+
+		dir := filepath.Dir(abs)
+
+		for {
+			gitignore := filepath.Join(dir, ".gitignore")
+			log.Printf("Checking .gitignore in %s\n", gitignore)
+			ignorePath, err := filepath.Rel(dir, abs)
+			pattern, lineNumber, err := findPatternInIgnoreFile(c, gitignore, ignorePath, isDir)
```

Comment:
What about matches that come from .gitignores higher up in the tree? I think the logic for directories is easier to reason about recursively than iteratively. For instance, we could parse the ignore file and return a slice of patterns, then pass that slice to children to deal with parent ignores. (Which also has the advantages of only parsing the file once). If that was done and the ls-files inversion, it would be easy to add exclude-per-directory and exclude-from

Reply:
Yes, I could think about this a bit more. Some of the complexity comes from the fact that in order to match a pattern in a gitignore file you need to make the path of the file to be ignore/not ignored relative to that gitignore file in order to check the matches.